### ISSUE TITLE
Incorrect Circular Dependency Detected w/ pred-constraints

### ISSUE THREAD

**Issue by rschmukler** [xtdb/xtdb#1523]
It seems the result binding vectors can trip up the query planners' dependency resolution. The following query (testable in `crux.lucene-test`) will break with a circular dependency exception.

```clojure
(t/deftest circular-test
  (submit+await-tx [[:crux.tx/put {:crux.db/id :ivan :name "Ivan" :surname "Evans"}]])
  (submit+await-tx [[:crux.tx/put {:crux.db/id :fred :name "Fred" :surname "Evans"}]])
  (submit+await-tx [[:crux.tx/put {:crux.db/id :matt :name "Matt" :surname "Stevens"}]])
  (with-open [db (c/open-db *api*)]
    (c/q db '{:find  [?a-name ?b-name]
              :where [[?a :name ?a-name]
                      [?a :surname ?lastname]
                      [(text-search :surname ?lastname) [[?b]]]
                      [?b :name ?b-name]]})
    (t/is true)))
```

```
clojure.lang.ExceptionInfo: Circular dependency between ?lastname and ?a {:reason :com.stuartsierra.dependency/circular-dependency, :node ?lastname, :dependency ?a}
```

**Comment by jarohen** [comment]
Thanks @rschmukler - I've added your breaking test to our repo. 

At first glance it seems that the initial join order isn't correctly taking the predicate function return into account - will first seek to understand a little more of the history around this, and will let you know!

James

**Comment by rschmukler** [comment]
Awesome thank you @jarohen  - I believe the query performance issue that I mentioned in slack (see below) may also be related to the same (or similar) join ordering bug:

```clojure
{:find [?a-name ?b-age]
 :limit 5
 :where [[_ :person/name ?a-name]
         [(text-search :person/name ?a-name) [[?b]]
         [?b :person/age ?b-age]]}
```

In a node w/ ~800K records, the above query will timeout as it performs the EAV tripples first and then attempts to filter on the predicate.

**Comment by jarohen** [comment]
Hey @rschmukler,

I now have a PR for this one (#1530) - it might be a one-liner but it's quite deep in the query engine, so a lot of nuances :smile:  Our tests are green but I think it's worth spending some more time checking it against different edge cases. In the meantime, I've deployed a `dev-SNAPSHOT` version - could you confirm whether this fixes your issue?

Cheers,

James

**Comment by rschmukler** [comment]
@jarohen Thanks for submitting #1530. I can confirm that `dev-SNAPSHOT` was indeed able to run the query above, as in, the crux/q function did not throw an error immediately. It now suffers from the same join order bug that I mentioned above where it attempts to run the predicate on all 800K records before unifying. Here is the query and debug output:

Edit: Looking at the logs below it looks like `?nearest` is being prioritized above `?person` which is likely why it has to run the predicate over every value in the database.

```clojure
{:find  [?name ?nearest-name]
 :limit 1
 :where [[?person :person/name ?name]
         [?person :person/location ?a-loc]
         [(geo-nearest :person/location ?a-loc) [[?nearest]]]
         [?nearest :person/name ?nearest-name]]}
```

```

debug 2021-06-10T23:22:01.489Z [crux.query:326] - :query {:find [?name ?nearest-name], :limit 1, :where [[?person :person/name ?name] [?person :person/location ?a-loc] [(geo-nearest :person/location ?a-loc) [[?nearest]]] [?nearest :person/name ?nearest-name]], :in []}
debug 2021-06-10T23:22:01.493Z [crux.query:326] - :triple-joins-var->cardinality {?a-loc 343409.9347944968, ?person 339786.2670920218, ?name 1.7976931348623157E308, ?nearest-name 1.7976931348623157E308, ?nearest 24.14356821632024}
debug 2021-06-10T23:22:01.493Z [crux.query:326] - :triple-joins-join-order [?nearest ?nearest-name ?person ?a-loc ?name]
debug 2021-06-10T23:22:01.496Z [crux.query:326] - :triple-joins-var->cardinality {?a-loc 343409.9347944968, ?person 339786.2670920218, ?name 1.7976931348623157E308, ?nearest-name 1.7976931348623157E308, ?nearest 24.14356821632024}
debug 2021-06-10T23:22:01.496Z [crux.query:326] - :triple-joins-join-order [?nearest ?nearest-name ?person ?a-loc ?name]
debug 2021-06-10T23:22:01.496Z [crux.query:326] - :project-only-leaf-vars #{?nearest-name ?name}
debug 2021-06-10T23:22:01.498Z [crux.query:326] - :join-order :aev ?person ?a-loc {:e ?person, :a :person/location, :v ?a-loc}
debug 2021-06-10T23:22:01.499Z [crux.query:326] - :join-order :aev ?person ?name {:e ?person, :a :person/name, :v ?name}
debug 2021-06-10T23:22:01.500Z [crux.query:326] - :join-order :aev ?nearest ?nearest-name {:e ?nearest, :a :person/name, :v ?nearest-name}
debug 2021-06-10T23:22:01.500Z [crux.query:326] - :where [[:triple {:e ?person, :a :person/name, :v ?name}] [:triple {:e ?person, :a :person/location, :v ?a-loc}] [:pred {:pred {:pred-fn geo-nearest, :args [:person/location ?a-loc]}, :return [:relation [[?nearest]]]}] [:triple {:e ?nearest, :a :person/name, :v ?nearest-name}]]
debug 2021-06-10T23:22:01.500Z [crux.query:326] - :vars-in-join-order [?nearest ?person ?a-loc cycle_?nearest_39126 ?nearest-name ?name]
debug 2021-06-10T23:22:01.500Z [crux.query:326] - :var->bindings {cycle_?nearest_39126 #crux.query.VarBinding{:e-var nil, :var cycle_?nearest_39126, :attr nil, :result-index 3, :result-name crux.query.value/cycle_?nearest_39126, :type :pred, :value? true}, ?person #crux.query.VarBinding{:e-var ?person, :var ?person, :attr :crux.db/id, :result-index 1, :result-name ?person, :type :entity, :value? false}, ?nearest #crux.query.VarBinding{:e-var ?nearest, :var ?nearest, :attr :crux.db/id, :result-index 0, :result-name ?nearest, :type :entity, :value? false}, ?a-loc #crux.query.VarBinding{:e-var ?person, :var ?a-loc, :attr :person/location, :result-index 2, :result-name ?person, :type :entity, :value? false}, ?name #crux.query.VarBinding{:e-var ?person, :var ?name, :attr :person/name, :result-index 5, :result-name ?person, :type :entity, :value? false}, ?nearest-name #crux.query.VarBinding{:e-var ?nearest, :var ?nearest-name, :attr :person/name, :result-index 4, :result-name ?nearest, :type :entity, :value? false}}

```

**Comment by refset** [comment]
This should be viable as a workaround:
```clj
{:find  [?name ?nearest-name]
 :limit 1
 :rules [[(get-nearest-name [?nearest] ?nearest-name)
          [?nearest :person/name ?nearest-name]]]
 :where [[?person :person/name ?name]
         [?person :person/location ?a-loc]
         [(geo-nearest :person/location ?a-loc) [[?nearest]]]
         (get-nearest-name ?nearest ?nearest-name)]}
```

Failing that, the built-in `get-attr` predicate or a subquery would also be options to explore.

**Comment by jarohen** [comment]
Fix for the original issue is in `"1.18.0-rc1"`
